### PR TITLE
New Recommender PNMF 'Projective Non-Negative Matrix Factorization' + Bugfix TextDataConvertor

### PR DESCRIPTION
**Pull request by dvelten**
Hello

I have made a new recommender PNMF - 'Projective Non-Negative Matrix Factorization'
There is no research paper analyzing the quality of this algorithm as recommender.
Research of the cited research papers was more general or was focused on image processing.
But I think it is a very good algorithm for recommender usecase too.
Algorithm is also fast on large sparse datasets because computation must only be done on items rated. 

There is also Bugfix for loading the movielens 20m dataset in class TextDataConvertor.
Code before was not able to load the dataset even when header-row of the ratings.csv was removed. Now it works, but header row must still be removed manually.

### CODE REVIEW

**Comment by guoguibing** [comment]
Thanks for your contribution. We are working on it.

**Comment by dvelten** [comment]
I have added now another Recommender: NMFItemItemRecommender

It is like PNMF but with second matrix must not be the transposed first matrix.

This seems to make the results better.

There are also other optimisation like in the fism paper. Do not estimate an item with itself in train phase.

**Comment by SunYatong** [comment]
@dvelten 
Thanks for your contribution. I created another pull request  #196  based on yours with some modifications to avoid conflicts with our new version. Please check my code in case I made mistakes. 
By the way, your documentation is great and I read the reference papers. I found that  PNMF and NMFItemItem are really similar to FISM, the main difference is the restrictions of learned parameter matrices :  NMFItemItem is the no-negative version of FISM and PNMF adds transpose restriction based on NMFItemItem. Am I right ?

**Comment by dvelten** [comment]
I have not checked the modified code now. I will check later and write again if something is wrong.

I assumed PNMF and NMFItemItem to be very different to FISM. Main difference is the loss functions (and bias handling). But it is right. FISM paper has shown that you can optimize an item/item matrix further. They worked exactly and discovered that others like NSVD and svd++ suffers from self estimation while training. So I have included this in NMFItemItem as configurable switch. You can test the effect yourself. There is also other optimization in FISM called 'neighborhood agreement'. I found that it is not working in my case. I think that square root is the right term there, because they use a quadratic loss.

I started working on PNMF and NMFItemItem because I wanted a Item to Item version of the Item to User version.

Item to Item is simply needed for fast online recommendations with lots of new users. So you do not want to store information about each user in the recommender model.

I wanted to use NMF with I-Divergence (or ~Kullback-Leibler divergence) loss function because it has nice property: Only calculation of bought items is done. This is just necessary on big data.

I even started with reading of 'Scalable Recommendation with Hierarchical Poisson Factorization' but found that HPoisson is too complicated to do the switch to an item to item recommender.

PNMF has transpose restriction like original paper.
NMFItemItem has this restriction removed. With experiments I found that this improves results already very well.
FISM Optimization is also very important! Without FISM Optimization 'Do not estimate yourself' results are bad especially on more sparse data.

But even with optimization I do not beat the BPR-User-Item-Recommender with same count of latent factor dimensions.

I currently do testing if Item-Item recommenders perhaps profit more from more latent factors. (Number of free parameters is lower with same count of latent factor dimensions.)

Perhaps I want to create Item-to Item version of BPR or compare with Item to User version of NMF for further investigation.

There is also other topic open for further research: Original NMF paper 'Learning the parts of an Object' in 'Nature' has found that NMF has the property to fraction the parts more semantically than other approaches. But there is only discussion like the User to Item Matrix. With the Item-to Latent Factor is more like a reconstruction matrix. But brain would use the 'Pseudoinverse' Analyzing Matrix like the second matrix in NMFItemItem for detection of the parts. But this second matrix is not discussed in original papers.. Perhaps Analyzing and Reconstruction Matrix can be compared visually like in original paper with images.

I am only working seldom on this topic. At company I am working usually on different topics. So do not know when I have further results.

**Comment by dvelten** [comment]
I have forgotten to mention that there is next problem: Nothing can be estimated in a complete natural way when no item is bought before or nothing is known about the user. Information of the user is only used for training if he bought at least 2 items.

This could be solved in different ways. 

Perhaps this is the cause that this approach still suffers against BPR?

Best wishes

Daniel Velten

**Comment by dvelten** [comment]
I think you hinted me to open problem again: Handling of unknown user.

FISMrmse is doing this with bias. I thought I can omit bias because it is usually only needed in Quadratic loss MF but not in NMF. But it has in this case also the functionality of handling the unknown user.

PseudoInvers second analyzing matrix could not even be a linear function of only the bought items simply because unkown user has nothing bought, but next bought item probability must be estimated as one.

So you need perhaps such trade-of as described in FISMrmse between unkown user estimate (the item bias) and the linear cross item recommendation.

I will try this later in year again.

**Comment by SunYatong** [comment]
@dvelten 
Or you can split the data yourself to make sure that every user has appeared in the training set. Then the recommenders can be compared on the known users only. 

BTW I'm looking forward for your result of combination of NMFItemItem and BPR. Maybe you can publish a paper if it gets significant improvement.

**Comment by SunYatong** [comment]
@dvelten 
Hi, may I have your name please? We 'd like to add you to our contributor list.

**Comment by dvelten** [comment]
Oh I am very glad, that you want to add me to the contributors list. My name is 'Daniel Velten'.

I want to further optimize the NMFItemItemRecommender. I could contribute this perhaps later this year.

I do work at private company 'econda GmbH' and do not want to publish scientific research paper. But could provide some documentation in form of latex/pdf document. Is there a place for such documentation in your repository?

**Comment by SunYatong** [comment]
@dvelten 
We haven't got any algorithms' detail documentation, I have to asking for Instructions from our leader.

By the way, I found that your implementation of NoveltyEvaluator is different from the reference paper "Solving the apparent diversity-accuracy dilemma of recommender systems." And I wish to consult you on the reason of your modification.

The process of paper is:
- calculate the self-information I_{alpha} for each Item.
- calculate the mean self-information of each user’s topN objects.
- averaging over users with at least one deleted link.

Your process is:
- calculate the self-information I_{alpha} for each Item.
- sum k_{alpha} * I_{alpha} for all items.
- divide the summation by (nonZeroNumUsers * Math.log(2))

**Comment by dvelten** [comment]
Sorry, 

You are right: This Implementation of NoveltyEvaluator is not this mentioned in the paper. But it is also a well known and often used Evaluator. But it is more a 'Diversity'-'Entropy' - Evaluator only. 

I have now created two Implementations. The complete new corrected NoveltyEvaluator and an EntropyEvaluator containing an modified version of the old NoveltyEvaluator.

Both Implementation use Entropy or Information to do calculations. The new EntropyEvaluator does not take into account if the item purchased is popular or unpopular in respect of probability purchased.

Instead of this the probability space is different. 

NoveltyEvaluator: Probability of item purchased
EntropyEvaluator: Probability of item to be in the recommender result list

Both Implementations could be thought to make the same for loops to sum the Entropy of one result list.

But in the EntropyEvaluator you can do an optimization and rewrite the for loop over users into an for loop over items with the pre-calculated item counts (item probabilities).

I make a new Pull-Request soon...

For documentation of PNNF there exists the paper... Surely it is not thought to be an recommender, but all calculations are there.

For NMFItemItem I will provide documentation and and further optimized version, but need longer time for doing this. Perhaps it will be next year.

Best wishes

Daniel
